File: backend/src/integrations/kms/kms_discovery_service.js
/**
 * KMS & HSM Discovery Service — Phase 14.3
 *
 * Central orchestration service managing KMS/Vault/HSM connectors.
 * Aggregates cryptographic key metadata fleetwide and exports to CycloneDX CBOM format.
 *
 * Invariants:
 * - Read-only discovery only.
 * - Zero private key extraction.
 */

const crypto = require("crypto");
const { KmsKeyMetadata } = require("./kms_metadata");
const { BaseKmsConnector } = require("./base_kms_connector");

class KmsDiscoveryService {
  constructor() {
    this.connectors = new Map();
  }

  /**
   * Registers a KMS connector instance.
   * @param {BaseKmsConnector} connector
   */
  registerConnector(connector) {
    if (!(connector instanceof BaseKmsConnector)) {
      throw new Error("Connector must inherit from BaseKmsConnector");
    }
    if (this.connectors.has(connector.name)) {
      throw new Error(`Connector with name '${connector.name}' already registered`);
    }
    this.connectors.set(connector.name, connector);
    return connector;
  }

  /**
   * Gets a registered connector by name.
   * @param {string} name
   */
  getConnector(name) {
    return this.connectors.get(name) || null;
  }

  /**
   * Unregisters a connector.
   * @param {string} name
   */
  unregisterConnector(name) {
    return this.connectors.delete(name);
  }

  /**
   * Lists all registered connectors and their metadata.
   */
  listConnectors() {
    return Array.from(this.connectors.values()).map((c) => ({
      name: c.name,
      provider: c.provider,
      readOnly: c.readOnly,
      config: c.sanitizeConfig(),
      leastPrivilegeRole: c.getLeastPrivilegeRoleDefinition(),
    }));
  }

  /**
   * Executes key discovery across all or filtered connectors.
   * @param {Object} [filter]
   * @param {string} [filter.provider] - Optional provider filter
   * @param {string} [filter.connectorName] - Optional specific connector
   * @returns {Promise<{ timestamp: string, totalCount: number, keys: KmsKeyMetadata[], errors: Object[] }>}
   */
  async discoverAll(filter = {}) {
    let targets = Array.from(this.connectors.values());

    if (filter.connectorName) {
      targets = targets.filter((c) => c.name === filter.connectorName);
    }
    if (filter.provider) {
      targets = targets.filter((c) => c.provider === filter.provider.toLowerCase());
    }

    const allKeys = [];
    const errors = [];

    for (const connector of targets) {
      try {
        const keys = await connector.discoverAll();
        allKeys.push(...keys);
      } catch (err) {
        errors.push({
          connectorName: connector.name,
          provider: connector.provider,
          error: err.message,
        });
      }
    }

    return {
      timestamp: new Date().toISOString(),
      totalCount: allKeys.length,
      keys: allKeys,
      errors,
    };
  }

  /**
   * Exports an array of KmsKeyMetadata into a CycloneDX Cryptographic Bill of Materials (CBOM).
   * @param {KmsKeyMetadata[]} keys
   * @returns {Object} CycloneDX CBOM JSON Document
   */
  toCbom(keys = []) {
    const keyList = Array.isArray(keys) ? keys : [];
    const components = keyList.map((k) =>
      k instanceof KmsKeyMetadata ? k.toCbomComponent() : new KmsKeyMetadata(k).toCbomComponent()
    );

    return {
      bomFormat: "CycloneDX",
      specVersion: "1.6",
      serialNumber: `urn:uuid:${crypto.randomUUID()}`,
      version: 1,
      metadata: {
        timestamp: new Date().toISOString(),
        tools: {
          components: [
            {
              type: "application",
              name: "ECDAT-KMS-Discovery-Engine",
              version: "1.0.0",
              vendor: "ECDAT",
            },
          ],
        },
        properties: [
          { name: "ecdat:module", value: "cloud_kms_hsm_metadata" },
          { name: "ecdat:extraction_policy", value: "metadata_only_read_only" },
        ],
      },
      components,
    };
  }

  /**
   * Generates a summary analysis of the discovered keys.
   * @param {KmsKeyMetadata[]} keys
   */
  getSummary(keys = []) {
    const list = Array.isArray(keys) ? keys : [];
    const byProvider = {};
    const byAlgorithm = {};
    const byState = {};

    let unrotatedCount = 0;
    let exportableCount = 0;
    let weakKeyCount = 0;

    for (const key of list) {
      // By provider
      byProvider[key.provider] = (byProvider[key.provider] || 0) + 1;

      // By algorithm
      byAlgorithm[key.algorithm] = (byAlgorithm[key.algorithm] || 0) + 1;

      // By state
      byState[key.state] = (byState[key.state] || 0) + 1;

      // Rotation check
      if (!key.rotation || !key.rotation.enabled) {
        unrotatedCount++;
      }

      // Exportability check
      if (key.usage && key.usage.isExportable) {
        exportableCount++;
      }

      // Weak key check (e.g. RSA < 2048, 3DES, DES)
      const numSize = typeof key.size === "number" ? key.size : parseInt(key.size, 10);
      if (
        (key.algorithm.includes("RSA") && numSize < 2048) ||
        key.algorithm.includes("3DES") ||
        key.algorithm.includes("DES")
      ) {
        weakKeyCount++;
      }
    }

    return {
      totalKeys: list.length,
      byProvider,
      byAlgorithm,
      byState,
      metrics: {
        unrotatedCount,
        exportableCount,
        weakKeyCount,
      },
    };
  }
}

// Singleton instance
const defaultKmsDiscoveryService = new KmsDiscoveryService();

module.exports = {
  KmsDiscoveryService,
  defaultKmsDiscoveryService,
};

File: scanners/integrations/kms/kms_discovery_service.py
"""
KMS & HSM Discovery Service (Python) — Phase 14.3

Registry, discovery orchestration, CBOM generation, and fleet analysis for KMS/HSM keys.
"""

from typing import Dict, Any, List, Optional
import uuid
import datetime
from .kms_metadata import KmsKeyMetadata
from .base_kms_connector import BaseKmsConnector


class KmsDiscoveryService:
    def __init__(self):
        self.connectors: Dict[str, BaseKmsConnector] = {}

    def register_connector(self, connector: BaseKmsConnector) -> BaseKmsConnector:
        if not isinstance(connector, BaseKmsConnector):
            raise TypeError("Connector must inherit from BaseKmsConnector")
        if connector.name in self.connectors:
            raise ValueError(f"Connector with name '{connector.name}' already registered")
        self.connectors[connector.name] = connector
        return connector

    def get_connector(self, name: str) -> Optional[BaseKmsConnector]:
        return self.connectors.get(name)

    def unregister_connector(self, name: str) -> bool:
        return self.connectors.pop(name, None) is not None

    def list_connectors(self) -> List[Dict[str, Any]]:
        return [
            {
                "name": c.name,
                "provider": c.provider,
                "read_only": c.read_only,
                "config": c.sanitize_config(),
                "least_privilege_role": c.get_least_privilege_role_definition(),
            }
            for c in self.connectors.values()
        ]

    def discover_all(self, filter_opts: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        filters = filter_opts or {}
        targets = list(self.connectors.values())

        if filters.get("connector_name"):
            targets = [c for c in targets if c.name == filters["connector_name"]]
        if filters.get("provider"):
            targets = [c for c in targets if c.provider == str(filters["provider"]).lower()]

        all_keys: List[KmsKeyMetadata] = []
        errors: List[Dict[str, Any]] = []

        for connector in targets:
            try:
                keys = connector.discover_all()
                all_keys.extend(keys)
            except Exception as e:
                errors.append(
                    {
                        "connector_name": connector.name,
                        "provider": connector.provider,
                        "error": str(e),
                    }
                )

        return {
            "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
            "total_count": len(all_keys),
            "keys": all_keys,
            "errors": errors,
        }

    def to_cbom(self, keys: Optional[List[Any]] = None) -> Dict[str, Any]:
        key_list = keys or []
        components = [
            k.to_cbom_component() if isinstance(k, KmsKeyMetadata) else KmsKeyMetadata(**k).to_cbom_component()
            for k in key_list
        ]

        return {
            "bomFormat": "CycloneDX",
            "specVersion": "1.6",
            "serialNumber": f"urn:uuid:{uuid.uuid4()}",
            "version": 1,
            "metadata": {
                "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
                "tools": {
                    "components": [
                        {
                            "type": "application",
                            "name": "ECDAT-KMS-Discovery-Engine",
                            "version": "1.0.0",
                            "vendor": "ECDAT",
                        }
                    ]
                },
                "properties": [
                    {"name": "ecdat:module", "value": "cloud_kms_hsm_metadata"},
                    {"name": "ecdat:extraction_policy", "value": "metadata_only_read_only"},
                ],
            },
            "components": components,
        }

    def get_summary(self, keys: Optional[List[KmsKeyMetadata]] = None) -> Dict[str, Any]:
        key_list = keys or []
        by_provider: Dict[str, int] = {}
        by_algorithm: Dict[str, int] = {}
        by_state: Dict[str, int] = {}

        unrotated_count = 0
        exportable_count = 0
        weak_key_count = 0

        for key in key_list:
            by_provider[key.provider] = by_provider.get(key.provider, 0) + 1
            by_algorithm[key.algorithm] = by_algorithm.get(key.algorithm, 0) + 1
            by_state[key.state] = by_state.get(key.state, 0) + 1

            if not key.rotation.get("enabled"):
                unrotated_count += 1

            if key.usage.get("is_exportable"):
                exportable_count += 1

            try:
                num_size = int(key.size)
            except Exception:
                num_size = 256

            if ("RSA" in key.algorithm and num_size < 2048) or "3DES" in key.algorithm or "DES" in key.algorithm:
                weak_key_count += 1

        return {
            "total_keys": len(key_list),
            "by_provider": by_provider,
            "by_algorithm": by_algorithm,
            "by_state": by_state,
            "metrics": {
                "unrotated_count": unrotated_count,
                "exportable_count": exportable_count,
                "weak_key_count": weak_key_count,
            },
        }


default_kms_discovery_service = KmsDiscoveryService()
